src/flatten.jl

```diff
@@ -21,20 +21,20 @@ flatten(x) = flatten(Float64, x)
 
 function flatten(::Type{T}, x::Integer) where {T<:Real}
     v = T[]
-    unflatten_to_Integer(v::Vector{T}) = x
+    unflatten_to_Integer(v::AbstractVector{<:Real}) = x
     return v, unflatten_to_Integer
 end
 
 function flatten(::Type{T}, x::R) where {T<:Real,R<:Real}
     v = T[x]
-    unflatten_to_Real(v::Vector{T}) = convert(R, only(v))
+    unflatten_to_Real(v::AbstractVector{<:Real}) = only(v)
     return v, unflatten_to_Real
 end
 
-flatten(::Type{T}, x::Vector{R}) where {T<:Real,R<:Real} = (Vector{T}(x), Vector{R})
+flatten(::Type{T}, x::Vector{R}) where {T<:Real,R<:Real} = (Vector{T}(x), identity)
 
 function _flatten_vector_integer(::Type{T}, x::AbstractVector{<:Integer}) where {T<:Real}
```

Comment:
Where is this function actually needed? Could we just do:
```julia
function flatten(::Type{T}, x::Union{I, Array{I}}) where {I<:Integer}
    v = T[]
    unflatten_to_Integer(v::Vector{T}) = x
    unflatten_to_Integer(v::AbstractVector{<:Real}) = x
    return v, unflatten_to_Integer
end
```
or do we get a method collision somewhere? (if yes we can change array to Vector)

Reply:
I think if we didn't do this we were hitting the wrong method for something. Might be possible to reduce the amount of code required as you suggest, but I'm going to leave for now.